### PR TITLE
[WIP] feat(docs): Add Falco and Fargate with ECS documentation

### PR DESCRIPTION
**Pull request by kris-nova**
Note: The Falco community will be updating our artifacts and documentation for these new features. This PR should be considered a WIP (Work in Progress) until then.

Signed-off-by: Kris Nova <kris@nivenly.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: doc_source/falco-fargate-ptrace-security.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                }
+            },
+```
+
+## Installing Falco
+
+_Note:_ The Falco `pdig` components are targetted to be released in an upcoming Falco release. This documentation will need to be updated after the initial release.
+
+For now you can use the `krisnova/falco-trace:latest` container image which is built on debian and contains Falco and `pdig` pre-installed. 
+The original work to get this working can be found at [github.com/kris-nova/falco-trace](https://github.com/kris-nova/falcotrace).
```
Comment (kris-nova) [inline doc_source/falco-fargate-ptrace-security.md:43]: These lines (40-44) will ultimately be removed before I remove the WIP flag from the PR.

But until the Falco community is able to cut a release with these updates, this is a place for folks to kick the tires.

File: doc_source/falco-fargate-ptrace-security.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+After Falco is logging to a file you can run your application.
+
+```bash
+pdig -a ./myapp &
+falco -u --daemon 
+tail -f /var/log/faclo.log &
+```
+
+The logs will now be picked up by ECR and can be used with other AWS services.
```
Comment (PettitWesley) [inline doc_source/falco-fargate-ptrace-security.md:119]: I don't think the logs will be picked up by ECR... this must be a typo...

Should probably add that you can use FireLens to send the stdout logs somewhere

Comment (nathanpeck) [reply doc_source/falco-fargate-ptrace-security.md:119]: It looks like Falco also supports:

```
stdout_output:
  enabled: true
```

If this is enabled then Falco would just output alerts to stdout directly and FireLens can be used to split the Falco logs out from the application logs and route them to a different alert log group or wherever the customer wants.

@kris-nova Is there a performance advantage or something else I'm missing here that makes it better to use file_output + tail vs just outputting directly to stdout?

File: doc_source/falco-fargate-ptrace-security.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  keep_alive: false
+  filename: /var/log/falco.log
+```
+
+After Falco is logging to a file you can run your application.
+
+```bash
+pdig -a ./myapp &
+falco -u --daemon 
+tail -f /var/log/faclo.log &
```
Comment (gnosek) [inline doc_source/falco-fargate-ptrace-security.md:116]:
```suggestion
tail -f /var/log/falco.log &
```

File: doc_source/falco-fargate-ptrace-security.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+In order to begin securing a task in ECS a few components need to be installed within the container.
+
+ - Falco userspace daemon
+ - Falco pdig tracing utility
+ 
+After these components are installed in the container, Falco can be used in the following ways.
+
+ - Launch an arbitrary process, and begin tracing the original process and all of it's child processes. 
+ - Attach to a running process, and begin tracing the process and all of it's child processes.
```
Comment (gnosek) [inline doc_source/falco-fargate-ptrace-security.md:21]:
```suggestion
 - Launch an arbitrary process, and begin tracing the original process and all of its child processes. 
 - Attach to a running process, and begin tracing the process and all of its child processes. 
```

(attaching to multiple threads is not implemented yet, so the second bullet is ~half true for now)

File: doc_source/falco-fargate-ptrace-security.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
+# Securing Fargate Tasks with Falco Runtime Security
+
+ECS Fargate `1.4.0` [announced support](https://aws.amazon.com/about-aws/whats-new/2020/04/aws-fargate-launches-platform-version-14/) for the `CAP_SYS_PTRACE` linux capability.
+
+This new feature can be implemented with [Falco](falco.org), an open source runtime security project originally built by Sysdig, Inc and later donated to the Cloud Native Computing Foundation.
```
Comment (leodido) [inline doc_source/falco-fargate-ptrace-security.md:5]:
```suggestion
This new feature can be implemented with [Falco](falco.org), an [open source runtime security project](https://github.com/falcosecurity/falco) originally built by Sysdig, Inc and later donated to the Cloud Native Computing Foundation.
```

I would personally link also the repository on GitHub for people willing to deep-dive into the code.

File: doc_source/falco-fargate-ptrace-security.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+Falco uses `ptrace(2)` with Fargate in ECS to detect anomalous behavior at runtime. 
+
+Falco can be configured to send alerts to STDOUT such that ECS can consume the logs. They can then be used to trigger alerts and alarms using AWS logging solutions such as CloudWatch. 
+
+## Dependencies
+
+In order to begin securing a task in ECS a few components need to be installed within the container.
+
+ - Falco userspace daemon
+ - Falco pdig tracing utility
```
Comment (leodido) [inline doc_source/falco-fargate-ptrace-security.md:16]:
```suggestion
 - Falco [pdig](https://github.com/falcosecurity/pdig) tracing utility
```

File: doc_source/falco-fargate-ptrace-security.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+Falco uses `ptrace(2)` with Fargate in ECS to detect anomalous behavior at runtime. 
+
+Falco can be configured to send alerts to STDOUT such that ECS can consume the logs. They can then be used to trigger alerts and alarms using AWS logging solutions such as CloudWatch. 
+
+## Dependencies
+
+In order to begin securing a task in ECS a few components need to be installed within the container.
+
+ - Falco userspace daemon
```
Comment (leodido) [inline doc_source/falco-fargate-ptrace-security.md:15]:
```suggestion
 - [Falco](https://github.com/falcosecurity/falco) userspace daemon
```

File: doc_source/falco-fargate-ptrace-security.md (hunk 4 above)
Comment (kris-nova) [reply doc_source/falco-fargate-ptrace-security.md:21]: Can we open a feature for this in the Falco `pdig` repository? Can you also suggest some language that would be more appropriate here?

Comment (kris-nova) [reply doc_source/falco-fargate-ptrace-security.md:21]: PS: Thank you for the typo suggestion - will accept once we figure out the threading things :heart:

File: doc_source/falco-fargate-ptrace-security.md (hunk 2 above)
Comment (kris-nova) [reply doc_source/falco-fargate-ptrace-security.md:119]: Hey everyone -

> I don't think the logs will be picked up by ECR... this must be a typo...

Yes. This is a typo :smiley_cat: will fix.

> Should probably add that you can use FireLens to send the stdout logs somewhere

I don't know what FireLens is, so I don't really understand the expectation we want to capture here? Can you suggest something to add to the PR?

---

Regarding `stdout_output.enabled: true` 

That is already [baked into the container image](
https://github.com/kris-nova/falco-trace/blob/master/etc/falco/falco.yaml#L123-L124) I used in the demo and for this writeup. 

I had to tail the logs to effectively break the interactive mode (or at least annoy anyone with a TTY) to get the logs to show up.

I don't know very much about logging in AWS so any help or guidance would be appreciated.

Comment (nathanpeck) [reply doc_source/falco-fargate-ptrace-security.md:119]: Gotcha. Once the updates are made in Falco upstream I'll contribute a sample FireLens task definition config for this. Basically FireLens is a little managed Fluentbit sidecar that runs in your task, and can route the container's logs to different locations based on characteristics of the logs. I think we can make it separate out your application's HTTP access logs and other stdout output from the Falco alert logs themselves, for easier auditing / paging etc. By default ECS will take all the stdout and put it into a single Cloudwatch log group which would make it a bit harder to find the alerts among everything else, but FireLens could split these logs out into two different Cloudwatch log groups for example.

Comment (PettitWesley) [reply doc_source/falco-fargate-ptrace-security.md:119]: Use this as a reference: https://sysdig.com/blog/multi-cluster-security-firelens/

File: doc_source/falco-fargate-ptrace-security.md (hunk 4 above)
Comment (gnosek) [reply doc_source/falco-fargate-ptrace-security.md:21]: We need to implement attaching to multiple threads in pdig anyway so feel free to open an issue for this :)

I really hope the limitation is temporary, but currently the second bullet should be more like:

> Attach to a running thread, and begin tracing that thread and all of its future child threads and processes

(there's no such thing as a child thread but this isn't really the place to dig into this)

**Comment by joelbrandenburg** [comment]
Is this PR still a WIP or have we reached a consensus on the content?

**Comment by joelbrandenburg** [comment]
Going to close this for now.

**Pull request closed without merging**